### PR TITLE
sanoid: mostly usable, problems with syncoid

### PR DESCRIPTION
**Pull request by otwieracz**
Sanoid is mostly usable right now.
Syncoid has problems with binary files location (expects specific
utilities like mbuffer and zfs to be present on the same location on
both local and remote hosts (what is problematic on NixOS) - 
however, it is usable.

### CODE REVIEW

File: nixos/modules/services/backup/sanoid.nix [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+              };
+              
+              autoprune = mkOption {
+                type = types.nullOr types.bool;
+                default = null;
+                description = ''
+                  Prune old snapshots automatically.
+                '';
+              };
+recursive = mkOption {
```
Comment (fpletz) [inline nixos/modules/services/backup/sanoid.nix]: Please fix the indention here. :)

File: pkgs/top-level/perl-packages.nix [hunk 2]
```diff
@@ -2031,6 +2031,20 @@ let self = _self // overrides; _self = with self; {
...
     };
   };
+  
+  ConfigIniFiles = buildPerlPackage rec {
+    name = "Config-IniFiles-2.38";
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/W/WA/WADG/${name}.tar.gz";
+      sha256 = "0cz5gz5qk729nrbmamz84i3335cc3vfzz8z450crrvzip4isdpgc";
+    };
+#    propagatedBuildInputs = [ MixinLinewise ];
```
Comment (fpletz) [inline pkgs/top-level/perl-packages.nix]: Why is this commented? Either remove to include or add another comment to explain the rationale.

File: nixos/modules/services/backup/sanoid.nix [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    ${optionalString (cfg.autosnap == true) "autosnap=yes"}
+    ${optionalString (cfg.autosnap == false) "autosnap=no"}
+    ${optionalString (cfg.autoprune == true) "autoprune=yes"}
+    ${optionalString (cfg.autoprune == false) "autoprune=no"}
+  '';
+in
+{
+  options = {
+    services.sanoid = {
+      enable = mkOption {
```
Comment (jagajaga) [inline nixos/modules/services/backup/sanoid.nix]: Please use `mkEnableOption`.

File: pkgs/tools/backup/sanoid/default.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    substituteInPlace $out/bin/syncoid --replace '/usr/bin/sudo' '/usr/bin/env sudo'
+    substituteInPlace $out/bin/syncoid --replace '/bin/ls' ${coreutils}/bin/ls
+    substituteInPlace $out/bin/syncoid --replace '/usr/bin/perl' ${perl}/bin/perl
+    substituteInPlace $out/bin/syncoid --replace '/bin/gzip' '/usr/bin/env gzip'
+    substituteInPlace $out/bin/syncoid --replace '/bin/zcat' '/usr/bin/env zcat'
+    substituteInPlace $out/bin/syncoid --replace '/usr/bin/lzop' '/usr/bin/env lzop'
+  '';
+
+  meta = {
+    description = "Policy-driven snapshot management and replication tools for ZFS.";
```
Comment (jagajaga) [inline pkgs/tools/backup/sanoid/default.nix]: Please remove dot at the end.

**Comment by zimbatm** [comment]
@otwieracz if you just squash those commits I think this would be good to go

**Comment by otwieracz** [comment]
Unfortunately, Sanoid is not able to work properly in NixOS environment.jj

**Comment by deliciouslytyped** [comment]
Not sure what happened here but current syncoid seems to have worked fine for what I just used it for. @otwieracz can you elaborate?

**Comment by otwieracz** [comment]
Sorry - this was more than two years ago and I am no longer using it.

**Comment by deliciouslytyped** [comment]
Do you remember the problem at the time?

**Comment by otwieracz** [comment]
Not really – only thing I remember is that it was expecting binary to be available in the same place on both sides of transfer. Or something like that. And it was by principle incompatibile with NixOS dynamic approach to binary location. --sg From: deliciouslytypedSent: wtorek, 13 sierpnia 2019 12:15To: NixOS/nixpkgsCc: Slawomir Gonet; MentionSubject: Re: [NixOS/nixpkgs] sanoid: mostly usable, problems with syncoid (#12609) 

 

   
  
 
  
  
    On August 13, 2019 at 12:15 PM deliciouslytyped <notifications@github.com> wrote:
   
    
   Do you remember the problem at the time?

**Comment by deliciouslytyped** [comment]
Ah. I only used it for a local transfer.

**Comment by deliciouslytyped** [comment]
What do you currently use?

**Comment by otwieracz** [comment]
I am no longer using ZFS snapshots as a backup solution at all.  --sg From: deliciouslytypedSent: wtorek, 13 sierpnia 2019 12:22To: NixOS/nixpkgsCc: Slawomir Gonet; MentionSubject: Re: [NixOS/nixpkgs] sanoid: mostly usable, problems with syncoid (#12609) 

 

   
  
 
  
  
    On August 13, 2019 at 12:22 PM deliciouslytyped <notifications@github.com> wrote:
   
    
   What do you currently use?

**Comment by lopsided98** [comment]
See #72060

**Pull request closed without merging**